File: MaxFlow.java
import java.util.*;

// Main class for the Maximum Flow problem
public class MaxFlow {

    // Method to find the maximum flow using Edmonds-Karp
    public static int edmondsKarp(int[][] capacity, int source, int sink) {
        int n = capacity.length; // Number of nodes in the graph
        int[][] residualCapacity = new int[n][n]; // Residual capacity graph
        for (int i = 0; i < n; i++) {
            System.arraycopy(capacity[i], 0, residualCapacity[i], 0, n);
        }

        int maxFlow = 0; // Variable to store the maximum flow

        // While there is an augmenting path from source to sink
        while (true) {
            int[] parent = new int[n]; // To store the path from source to sink
            Arrays.fill(parent, -1);
            boolean foundPath = bfs(residualCapacity, source, sink, parent);

            if (!foundPath) break; // No more augmenting paths, break the loop

            // Find the minimum capacity in the augmenting path
            int flow = Integer.MAX_VALUE;
            int current = sink;
            while (current != source) {
                int prev = parent[current];
                flow = Math.min(flow, residualCapacity[prev][current]);
                current = prev;
            }

            // Update the residual capacities of the edges and reverse edges
            current = sink;
            while (current != source) {
                int prev = parent[current];
                residualCapacity[prev][current] -= flow; // Reduce capacity along the path
                residualCapacity[current][prev] += flow; // Increase reverse capacity
                current = prev;
            }

            maxFlow += flow; // Add the flow of this path to the total flow
        }

        return maxFlow;
    }

    // Breadth-First Search to find an augmenting path
    private static boolean bfs(int[][] residualCapacity, int source, int sink, int[] parent) {
        int n = residualCapacity.length;
        boolean[] visited = new boolean[n]; // Track visited nodes
        Queue<Integer> queue = new LinkedList<>();
        queue.add(source);
        visited[source] = true;

        while (!queue.isEmpty()) {
            int u = queue.poll();

            // Explore all neighbors of u
            for (int v = 0; v < n; v++) {
                // If there is residual capacity and v is not visited
                if (!visited[v] && residualCapacity[u][v] > 0) {
                    parent[v] = u; // Store the path
                    visited[v] = true;
                    if (v == sink) return true; // Path to sink found
                    queue.add(v);
                }
            }
        }

        return false; // No augmenting path found
    }

    // Main method to test the algorithm
    public static void main(String[] args) {
        // Example graph (capacity matrix)
        int[][] capacity = {
            {0, 10, 10, 0, 0, 0}, // Node 0 (source)
            {0, 0, 2, 4, 8, 0},   // Node 1
            {0, 0, 0, 0, 9, 0},   // Node 2
            {0, 0, 0, 0, 0, 10},  // Node 3
            {0, 0, 0, 6, 0, 10},  // Node 4
            {0, 0, 0, 0, 0, 0}    // Node 5 (sink)
        };

        int source = 0; // Source node
        int sink = 5;   // Sink node

        int maxFlow = edmondsKarp(capacity, source, sink);
        System.out.println("Maximum Flow: " + maxFlow);
    }
}

File: MaxFlow.cc
#include <iostream>
#include <vector>
#include <queue>
#include <cstring>
#include <algorithm>
using namespace std;

const int MAXN = 500; // Maximum number of nodes
const int INF = 1e9;

vector<int> adj[MAXN];  // Adjacency list
int capacity[MAXN][MAXN]; // Capacity matrix
int parent[MAXN];  // Parent array for BFS

// BFS to find an augmenting path
bool bfs(int s, int t) {
    memset(parent, -1, sizeof(parent));
    parent[s] = -2; // Mark the source as visited
    queue<pair<int, int>> q;
    q.push({s, INF});

    while (!q.empty()) {
        int cur = q.front().first;
        int flow = q.front().second;
        q.pop();

        for (int next : adj[cur]) {
            if (parent[next] == -1 && capacity[cur][next] > 0) {
                // If the next node is unvisited and has capacity left
                parent[next] = cur;
                int new_flow = min(flow, capacity[cur][next]);
                if (next == t) return true; // Found an augmenting path
                q.push({next, new_flow});
            }
        }
    }
    return false;
}

// Edmonds-Karp: Maximum Flow Algorithm
int maxFlow(int s, int t) {
    int flow = 0;

    while (bfs(s, t)) {
        int cur = t;
        int path_flow = INF;

        // Find the bottleneck capacity in the augmenting path
        while (cur != s) {
            int prev = parent[cur];
            path_flow = min(path_flow, capacity[prev][cur]);
            cur = prev;
        }

        // Update capacities along the path
        cur = t;
        while (cur != s) {
            int prev = parent[cur];
            capacity[prev][cur] -= path_flow;
            capacity[cur][prev] += path_flow;
            cur = prev;
        }

        flow += path_flow;
    }

    return flow;
}

int main() {
    int n, k, c; // Number of doctors, vacation periods, max days per doctor
    cin >> n >> k >> c;

    vector<vector<int>> Dj(k); // Vacation days for each period
    int total_days = 0;

    // Input vacation days
    for (int i = 0; i < k; i++) {
        int m; // Number of days in this vacation period
        cin >> m;
        total_days += m;
        Dj[i].resize(m);
        for (int j = 0; j < m; j++) {
            cin >> Dj[i][j];
        }
    }

    int N = total_days; // Total number of vacation days
    int s = 0, t = n + N + 1; // Source and sink nodes

    // Input doctor availability
    for (int i = 1; i <= n; i++) {
        int m; // Number of days this doctor is available
        cin >> m;
        for (int j = 0; j < m; j++) {
            int day;
            cin >> day;
            adj[i].push_back(n + day);
            adj[n + day].push_back(i);
            capacity[i][n + day] = 1;
        }
        adj[s].push_back(i);
        adj[i].push_back(s);
        capacity[s][i] = c;
    }

    // Connect vacation days to sink
    for (int day = 1; day <= N; day++) {
        adj[n + day].push_back(t);
        adj[t].push_back(n + day);
        capacity[n + day][t] = 1;
    }

    // Calculate maximum flow
    int flow = maxFlow(s, t);

    if (flow == N) {
        cout << "Yes\n";
    } else {
        cout << "No\n";
    }

    return 0;
}
